source/common/common/posix/thread_impl.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <functional>
+
+#include "envoy/thread/thread.h"
+
+namespace Envoy {
+namespace Thread {
+
+/**
+ * Wrapper for a pthread thread. We don't use std::thread because it eats exceptions and leads to
+ * unusable stack traces.
+ */
+class ThreadImplPosix : public Thread {
+public:
+  ThreadImplPosix(std::function<void()> thread_routine);
```

Comment:
FWIW I suspect you don't need ThreadImplPosix to be exposed in the header -- it can be private to the ThreadFactoryImplPosix impl and in the cc file. This is not a big deal but I feel like it's nice to minimize what has to go into h files, for compile-speed and for human understanding of interfaces.

Reply:
I agree that we don't need ThreadImplPosix exposed in the header, but we do need ThreadImplWin32 exposed (so one of our Windows-specific classes can access data we don't want to put on the interface). It may be more consistent to have both exposed rather than one exposed and one placed in the cc file

Reply:
OK leave it exposed then, this is not a big deal, and if we do subclass Apple vs Linux in the future we'll need the shared impl exposed at least as a class declaration.